Support multiple devices

amaisano opened the pull request:
I own more than 1 flag device (the standard Flag, and the Bluetooth Flag). They both have the same product ids.

This just adds loops/lists to the mix, and applies the same code to every matching device. Tested with Flag and Bluetooth devices.

Note: Patterns are buggy on the bluetooth device.

amaisano committed:
Work with multiple devices

I own more than 1 flag device (the standard Flag, and the Bluetooth Flag). They both have the same product ids.

This just adds loops/lists to the mix, and applies the same code to every matching device. Tested with Flag and Bluetooth devices.

Note: Patterns are buggy on the bluetooth device.

vmitchell85 commented:
@amaisano first off thanks for contributing!

I was going to approve this (with minor changes requested) then I thought about it some more. Do you know if there is any way to differentiate multiple devices? I'm asking because I'm not sure all users would want all their devices set together.

amaisano commented:
Hi @vmitchell85 !

Hmm. Well you could accept an argument for "index" which would be `DEVICES[i]` so that you can ask `luxafor-linux.py color -i 0 ....` or `luxafor-linux.py color -i 1 ....` to write values. Without an -i value we default to all devices?

(I know I need some cleanup here - like removing logging 😅 )

The only different between these devices are the bus address and the manufacturer string (one is "Greynut" the other is something else... "Micro" something). Otherwise _all_ property values are identical.

vmitchell85 commented:
@amaisano Unfortunately I only have one luxafor so I cannot test to see what all information would be available. Especially differences between the USB and bluetooth versions. Not sure I can justify purchasing a bluetooth version just for this PR 😆 .

It seems based on your code that the `usb.core.find` call would come back with an array of items if there are more than one. I think checking the length of devices, and applying it using the `i` index you suggested and even adding a `-all` flag would be good.

amaisano committed:
Ability to address different devices

- If no `-d` argument is set or `-d` is set to `0`, default to writing the values to all connected Luxafor devices.
- Device "id" is 1-based, since we made the default `0`. Device "1" is the first device found. Sorry, all props on these USB devices are basically useless for identification.
- Used a lightweight library to run `writeValue()` twice in a sub-routine.

amaisano commented:
@vmitchell85 LMK what you think - now users can choose the device for each operation if they like. So far after multiple reboots, the order of my two luxafor devices have remained the same.

amaisano commented:
@vmitchell85 bumping this. Hope you are well!

vmitchell85 commented:
@amaisano sorry for the delay. The only thing I was still contemplating is whether we should be pulling in a library for the repeat. If it was a large codebase where it was needed in many places that would make sense, but for this little script I don't think it's worth it.

Can you remove that library and put the "double call" back in. You do that and I'll merge.

Thanks,
- Vince

amaisano commented on luxafor-linux.py:
```diff
@@ -4,7 +4,9 @@
 import argparse
 
-DEVICE = None
+from itertools import repeat
```
```suggestion```

amaisano committed:
avoid 3rd-party dependency

vmitchell85 commented:
I'll try to merge tonight

amaisano commented:
Thanks @vmitchell85 ! Just realized one of us should update the docs (Readme). You want to or should I?

vmitchell85 commented:
@amaisano Can you create another PR for it?

amaisano commented:
@vmitchell85  https://github.com/vmitchell85/luxafor-python/pull/15
